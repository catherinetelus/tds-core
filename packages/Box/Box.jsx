import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import { media } from '@tds/core-responsive'
import safeRest from '../../shared/utils/safeRest'

const isObject = item => {
  return item && typeof item === 'object' && !Array.isArray(item) && item !== null
}

const mergeDeep = (target, source) => {
  if (isObject(target) && isObject(source)) {
    Object.keys(source).forEach(key => {
      if (isObject(source[key])) {
        if (!target[key]) Object.assign(target, { [key]: {} })
        mergeDeep(target[key], source[key])
      } else {
        Object.assign(target, { [key]: source[key] })
      }
    })
  }
  return target
}

const spacing = {
  mobile: {
    1: '0.25rem',
    2: '0.5rem',
    3: '1rem',
    4: '1.5rem',
    5: '2rem',
    6: '2.5rem',
    7: '3rem',
    8: '4rem',
  },
  desktop: {
    1: '0.25rem',
    2: '0.5rem',
    3: '1rem',
    4: '2rem',
    5: '3rem',
    6: '4rem',
    7: '4.5rem',
    8: '6rem',
  },
}

const boxSpacing = (level, f) => {
  const mobileStyle = f(spacing.mobile[level])
  if (spacing.mobile[level] === spacing.desktop[level]) {
    return mobileStyle
  }
  const desktopStyle = media.from('md')(f(spacing.desktop[level]))

  return mergeDeep(mobileStyle, desktopStyle)
}

const flexDirectionStyles = ({ inline }) => ({ flexDirection: inline ? 'row' : 'column' })
const betweenStyles = ({ between, inline }) => {
  if (between == null) {
    return undefined
  }

  if (between === 'space-between') {
    return { justifyContent: 'space-between' }
  }

  const marginDirection = inline ? 'marginRight' : 'marginBottom'

  return boxSpacing(between, s => ({
    '> *:not(:last-child)': {
      [marginDirection]: s,
    },
  }))
}

const horizontalStyles = ({ horizontal }) => {
  if (horizontal == null) {
    return undefined
  }
  return boxSpacing(horizontal, s => ({
    paddingLeft: s,
    paddingRight: s,
  }))
}

const verticalStyles = ({ vertical }) => {
  if (vertical == null) {
    return undefined
  }
  return boxSpacing(vertical, s => ({
    paddingTop: s,
    paddingBottom: s,
  }))
}

const insetStyles = ({ inset }) => {
  if (inset == null) {
    return undefined
  }
  const vertical = verticalStyles({ vertical: inset })
  const horizontal = horizontalStyles({ horizontal: inset })

  return mergeDeep(vertical, horizontal)
}

const belowStyles = ({ below }) => {
  if (below == null) {
    return undefined
  }

  return boxSpacing(below, s => ({
    marginBottom: s,
  }))
}

const StyledBox = styled.div(props => ({
  display: props.between ? 'flex' : 'block',
  ...flexDirectionStyles(props),
  ...betweenStyles(props),
  ...horizontalStyles(props),
  ...verticalStyles(props),
  ...insetStyles(props),
  ...belowStyles(props),
}))

/**
 * Apply spacing within or around components.
 *
 * @version ./package.json
 */
const Box = ({ dangerouslyAddClassName, tag, ...rest }) => (
  <StyledBox {...safeRest(rest)} as={tag} className={dangerouslyAddClassName} />
)

Box.propTypes = {
  /**
   * Specify an HTML element to render, such as `section`.
   */
  tag: PropTypes.string,
  /**
   * Indent content from the container's top and bottom edge by applying padding.
   */
  vertical: PropTypes.oneOf([1, 2, 3, 4, 5, 6, 7, 8]),
  /**
   * Indent content from the container's left and right edge by applying padding.
   */
  horizontal: PropTypes.oneOf([1, 2, 3, 4, 5, 6, 7, 8]),
  /**
   * Indent content from all of the container's edges by applying padding.
   */
  inset: PropTypes.oneOf([1, 2, 3, 4, 5, 6, 7, 8]),
  /**
   * @ignore
   *
   * We are keeping this hidden for now as we are not sold on the necessity. We use it internally still to apply
   * spacing to Markdown components, but would like to use between instead if the library allows it.
   *
   * Sets a `margin-bottom`.
   */
  below: PropTypes.oneOf([1, 2, 3, 4, 5, 6, 7, 8]),
  /**
   * Create either a block or an inline stack, applying margin in between every direct child. Margin will not be
   * applied to the last component in the stack.
   *
   * By default, `between` will arrange the Box's children as a flex column. Combine with `inline` to arrange them
   * as a flex row.
   */
  between: PropTypes.oneOf([1, 2, 3, 4, 5, 6, 7, 8]),
  /**
   * Arrange children in a row. Combine with `between` to apply margins in between the row's elements.
   */
  inline: PropTypes.bool,
  /**
   * Append custom classes to `className`. Use sparingly, and do not attempt to override Box style properties as that
   * may cause unexpected behaviour.
   *
   * You would typically use this feature to apply flex alignment properties in combination with `between`.
   */
  dangerouslyAddClassName: PropTypes.string,
  /**
   * The content. Can be text, any HTML element, or any component.
   */
  children: PropTypes.node.isRequired,
}

Box.defaultProps = {
  inline: false,
  tag: 'div',
  vertical: undefined,
  horizontal: undefined,
  inset: undefined,
  below: undefined,
  between: undefined,
  dangerouslyAddClassName: undefined,
}

export default Box
