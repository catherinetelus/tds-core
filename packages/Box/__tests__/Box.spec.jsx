import React from 'react'
import { shallow, mount } from 'enzyme'

import Box from '../Box'

describe('Box', () => {
  const defaultProps = { between: 3 }
  const doMount = (props = {}) => {
    const box = mount(
      <Box {...defaultProps} {...props}>
        Some content
      </Box>
    )
    return {
      box,
      styledComponent: box.find('StyledComponent'),
    }
  }
  const doShallow = (props = {}) => {
    return shallow(
      <Box {...defaultProps} {...props}>
        Some content
      </Box>
    )
  }

  it('renders', () => {
    const box = doMount()

    expect(box).toMatchSnapshot()
  })

  it('can render as a specified HTML element', () => {
    const box = doShallow({ tag: 'ul' })
    expect(box.props().as).toBe('ul')
  })

  it('can apply bottom margin', () => {
    const { styledComponent } = doMount({ below: 2 })
    expect(styledComponent).toHaveStyleRule('margin-bottom', '0.5rem')
  })

  describe('insets', () => {
    it('can be equal on all sides', () => {
      const { styledComponent } = doMount({ inset: 3 })

      expect(styledComponent).toHaveStyleRule('padding-top', '1rem')
      expect(styledComponent).toHaveStyleRule('padding-bottom', '1rem')
      expect(styledComponent).toHaveStyleRule('padding-left', '1rem')
      expect(styledComponent).toHaveStyleRule('padding-right', '1rem')
    })

    it('can be either vertical or horizonal', () => {
      let { styledComponent } = doMount({ vertical: 1 })
      expect(styledComponent).toHaveStyleRule('padding-top', '0.25rem')
      expect(styledComponent).toHaveStyleRule('padding-bottom', '0.25rem')

      styledComponent = doMount({ horizontal: 1 }).styledComponent
      expect(styledComponent).toHaveStyleRule('padding-left', '0.25rem')
      expect(styledComponent).toHaveStyleRule('padding-right', '0.25rem')

      styledComponent = doMount({ vertical: 2, horizontal: 3 }).styledComponent
      expect(styledComponent).toHaveStyleRule('padding-top', '0.5rem')
      expect(styledComponent).toHaveStyleRule('padding-bottom', '0.5rem')
      expect(styledComponent).toHaveStyleRule('padding-left', '1rem')
      expect(styledComponent).toHaveStyleRule('padding-right', '1rem')
    })
  })

  describe('between', () => {
    it('arranges the children horizontally or vertically be either inline or block', () => {
      let { styledComponent } = doMount()
      expect(styledComponent).toHaveStyleRule('display', 'flex')
      expect(styledComponent).toHaveStyleRule('flex-direction', 'column')

      styledComponent = doMount({ inline: true }).styledComponent
      expect(styledComponent).toHaveStyleRule('display', 'flex')
      expect(styledComponent).toHaveStyleRule('flex-direction', 'row')
    })

    it('separates children by equal margins in a stack', () => {
      const { styledComponent } = doMount({ between: 2 })

      expect(styledComponent).toHaveStyleRule('margin-bottom', '0.5rem', {
        modifier: '> *:not(:last-child)',
      })
    })

    it('separates children by equal margins inline', () => {
      const { styledComponent } = doMount({ between: 2, inline: true })

      expect(styledComponent).toHaveStyleRule('margin-right', '0.5rem', {
        modifier: '> *:not(:last-child)',
      })
    })
  })

  it('passes additional attributes to the wrapping element', () => {
    const box = doShallow({ id: 'the-id', 'data-some-attr': 'some value' })

    expect(box).toHaveProp('id', 'the-id')
    expect(box).toHaveProp('data-some-attr', 'some value')
  })

  it('will add additional arbitrary class names', () => {
    const box = doShallow({ dangerouslyAddClassName: 'a-class' })

    expect(box).toHaveClassName('a-class')
  })

  it('does not allow custom CSS', () => {
    const box = doShallow({ className: 'my-custom-class', style: { color: 'hotpink' } })

    expect(box).not.toHaveProp('className', 'my-custom-class')
    expect(box).not.toHaveProp('style')
  })
})
