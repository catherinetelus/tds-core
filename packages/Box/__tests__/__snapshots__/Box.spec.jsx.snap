// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Box renders 1`] = `
Object {
  "box": .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 > *:not(:last-child) {
  margin-bottom: 1rem;
}

<Box
    between={3}
    inline={false}
    tag="div"
  >
    <Box__StyledBox
      as="div"
      between={3}
      inline={false}
    >
      <StyledComponent
        as="div"
        between={3}
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "Box__StyledBox-jip4n1-0",
              "isStatic": false,
              "lastClassName": "c0",
              "rules": Array [
                [Function],
              ],
            },
            "displayName": "Box__StyledBox",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "Box__StyledBox-jip4n1-0",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
        inline={false}
      >
        <div
          className="c0"
        >
          Some content
        </div>
      </StyledComponent>
    </Box__StyledBox>
  </Box>,
  "styledComponent": .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 > *:not(:last-child) {
  margin-bottom: 1rem;
}

<StyledComponent
    as="div"
    between={3}
    forwardedComponent={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "Box__StyledBox-jip4n1-0",
          "isStatic": false,
          "lastClassName": "c0",
          "rules": Array [
            [Function],
          ],
        },
        "displayName": "Box__StyledBox",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "Box__StyledBox-jip4n1-0",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      }
    }
    forwardedRef={null}
    inline={false}
  >
    <div
      className="c0"
    >
      Some content
    </div>
  </StyledComponent>,
}
`;
